pkg/util/jsonpath/jsonpath.go

```
@@ -0,0 +1,153 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jsonpath
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"reflect"
+	"strconv"
+
+	"github.com/golang/glog"
+)
+
+type JSONPath struct {
+	name   string
+	parser *Parser
+}
+
+func NewJSONPath(n string) *JSONPath {
+	return &JSONPath{
+		name: n,
+	}
+}
+
+func (j *JSONPath) Parse(text string) (err error) {
+	j.parser, err = Parse(j.name, text)
+	return
+}
+
+func (j *JSONPath) Execute(wr io.Writer, data interface{}) error {
+	value := reflect.ValueOf(data)
+	if j.parser == nil {
+		return fmt.Errorf("%s is an incomplete jsonpath template", j.name)
+	}
+	j.walkTree(wr, value)
+	return nil
+}
+
+// walkTree visit the parsed tree from root
+func (j *JSONPath) walkTree(wr io.Writer, value reflect.Value) {
+	for _, node := range j.parser.Root.Nodes {
+		j.walk(wr, value, node)
+	}
+}
+
+// walk visit subtree rooted at the gived node in DFS order
+func (j *JSONPath) walk(wr io.Writer, value reflect.Value, node Node) reflect.Value {
+	switch node := node.(type) {
+	case *ListNode:
+		return j.evalList(wr, value, node)
+	case *TextNode:
+		return j.evalText(wr, value, node.Text)
+	case *FieldNode:
+		return j.evalField(value, node.Value)
+	case *ArrayNode:
+		return j.evalArray(value, node)
+	}
+	return reflect.Value{}
+}
+
+// evalText evaluate TextNode
+func (j *JSONPath) evalText(wr io.Writer, value reflect.Value, text []byte) reflect.Value {
+	if _, err := wr.Write(text); err != nil {
+		glog.Errorf("%s", err)
```

don't eat the error, make this method return err, and push it up the stack.